src/tests/ftest/cart/test_group_np_srv.c

```diff
@@ -43,17 +43,14 @@ swim_crt_event_cb(d_rank_t rank, enum crt_event_source src,
 		"rank = %d, "
 		"crt_event_source = %d, "
 		"crt_event_type = %d\n",
-		 rank, src, type);
+		rank, src, type);
 
-	/* convert integer to string */
-	snprintf(type_to_a, 2, "%d", type);
+	swim_state_str[0] = type + '0';
+	swim_state_str[1] = 0;
 
-	if (swim_status_by_rank[rank] == NULL) {
-		swim_status_by_rank[rank] = (char *)malloc(max * sizeof(char));
-		strcpy(swim_status_by_rank[rank], "\0");
-	}
-
-	strcat(swim_status_by_rank[rank], type_to_a);
+	maxlen = MAX_SWIM_STATUSES - strlen(swim_state_str);
+	if (strlen(swim_seq_by_rank[rank]) < maxlen)
```

Comment:
I'm not sure I grok what we are doing here

maxlen is a constant (MAX_SWIM_STATUSES - 1)

This check is checking to see if the existing string is shorter than that?

do you not want something like
if (strnlen(swim_seq_by_rank[rank], MAX_SWIM_STATUSES - 1) < (MAX_SWIM_STATUSES - 1))
    strcat(swim_seq_by_rank[rank], swim_state_str);
    
Or perhaps you actually want to reserve a character or more to save the fact that it's truncated?

Reply:
> I'm not sure I grok what we are doing here
> 
> maxlen is a constant (MAX_SWIM_STATUSES - 1)
> 
> This check is checking to see if the existing string is shorter than that?
> 
> do you not want something like
> if (strnlen(swim_seq_by_rank[rank], MAX_SWIM_STATUSES - 1) < (MAX_SWIM_STATUSES - 1))
> strcat(swim_seq_by_rank[rank], swim_state_str);

Works for me, but what if `strlen(swim_seq_by_rank[rank]) > maxlen` (using `n`-less `strlen`)?  Do we want to allow the `swim_seq_by_rank` buffer to overflow?

> Or perhaps you actually want to reserve a character or more to save the fact that it's truncated?

I wasn't marking the array as truncated, but should I?  Do want a special character array terminator to indicate the string is chopped?

Reply:
> > I'm not sure I grok what we are doing here
> > maxlen is a constant (MAX_SWIM_STATUSES - 1)
> > This check is checking to see if the existing string is shorter than that?
> > do you not want something like
> > if (strnlen(swim_seq_by_rank[rank], MAX_SWIM_STATUSES - 1) < (MAX_SWIM_STATUSES - 1))
> > strcat(swim_seq_by_rank[rank], swim_state_str);
> 
> Works for me, but what if `strlen(swim_seq_by_rank[rank]) > maxlen` (using `n`-less `strlen`)? Do we want to allow the `swim_seq_by_rank` buffer to overflow?

It won't overflow because we don't cat the string in that case.

> 
> > Or perhaps you actually want to reserve a character or more to save the fact that it's truncated?
> 
> I wasn't marking the array as truncated, but should I? Do want a special character array terminator to indicate the string is chopped?

I guess it depends on the preference for the test.   Since it's just a test, I suppose you control the environment somewhat and can avoid running a case that would overflow.   But it might be nice to give some sort of indication that it overflowed, either by using a special character to mark it or by printing something to indicate it happened.

Reply:
> > I wasn't marking the array as truncated, but should I? Do want a special character array terminator to indicate the string is chopped?
> 
> I guess it depends on the preference for the test. Since it's just a test, I suppose you control the environment somewhat and can avoid running a case that would overflow. But it might be nice to give some sort of indication that it overflowed, either by using a special character to mark it or by printing something to indicate it happened.

As the test runs now, I've only seen up to 3 of the 1024 allocated states used, e.g.,
```
SWIM state sequence (001) for status [1] is as expected.
```
Hitting 1024 swim states would seem to indicate a major change in swim behavior, but as the test is written now -- such a change could go under the radar.  Also maybe `MAX_SWIM_STATUSES` should be decreased, as it gives the impression that the test might reach that high number of states.